runtime/src/bank_forks.rs

```diff
@@ -86,8 +86,55 @@ impl Index<u64> for BankForks {
 
 impl BankForks {
     pub fn new_rw_arc(bank: Bank) -> Arc<RwLock<Self>> {
+        let bank = Arc::new(bank);
         let root = bank.slot();
-        Self::new_from_banks(&[Arc::new(bank)], root)
+
+        let mut banks = HashMap::new();
+        banks.insert(
+            bank.slot(),
+            BankWithScheduler::new_without_scheduler(bank.clone()),
+        );
+        let parents = bank.parents();
+        for parent in parents {
+            if banks
+                .insert(
+                    parent.slot(),
+                    BankWithScheduler::new_without_scheduler(parent.clone()),
+                )
+                .is_some()
+            {
+                // All ancestors have already been inserted by another fork
+                break;
+            }
+        }
+
+        let mut descendants = HashMap::<_, HashSet<_>>::new();
+        for (slot, bank) in &banks {
+            descendants.entry(*slot).or_default();
+            for parent in bank.proper_ancestors() {
+                descendants.entry(parent).or_default().insert(*slot);
+            }
+        }
```

Comment:
This is a holdover from the old code; we know that `banks` will have only one entry so we could just operate on that slot directly; thoughts?

Reply:
🤔 with this change should we even support banks that have parents & descendents? In the production cases, bank forks will always be initialized from a single bank that has no parent nor descendents.

Kinda hesitent to add an assert or make this return a result of some kind.

Reply:
There are the `ancestors` / `parent_slot` fields that are serialized. So, even if we only have a `Bank` for some snapshot slot, we may still be able to know what some of its' ancestors were. Also, consider the following two sequences of operations.

Sequence 1:
- Create `bank1`
- Create `BankForks` from `bank1`
- Create `bank2` via `new_from_parent()` on `bank1` and insert into `BankForks`

Sequence 2:
- Create `bank1`
- Create `bank2` via `new_from_parent()` on `bank1`
- Create `BankForks` from `bank1`
- Insert `bank2` into `BankForks`

`solana-validator` behavior matches sequence 1; however, I believe sequence 2 is technically correct and would be an example of a case where we'd have a descendant.

While it could seem a little "off" to have a descendant relation `bank2 descended from bank1` in the descendants structure without actually having `bank2` present, I believe this is the same as noting the root bank's parent without actually having the bank. That is, at `BankForks` creation, I think we'll note `bank1 descended from bank0` even if `bank1` is the root bank and we don't actually have `bank0`.

I can test this empirically and report back

Reply:
Just tested this, and confirmed that we do store some ancestor slots in snapshots. I added a quick log statement and restart my node; it restart from this slot:
```
[2023-12-14T03:55:23.085793677Z INFO  solana_runtime::snapshot_bank_utils] Loading bank from snapshot dir: /home/sol/ledger/snapshot/235777401
```
And I see all of these from the log statement I added:
```
[2023-12-14T03:56:07.311626674Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777368
[2023-12-14T03:56:07.311630822Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777369
[2023-12-14T03:56:07.311632715Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777370
[2023-12-14T03:56:07.311634709Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777371
[2023-12-14T03:56:07.311636543Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777372
[2023-12-14T03:56:07.311638266Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777373
[2023-12-14T03:56:07.311640059Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777374
[2023-12-14T03:56:07.311642053Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777375
[2023-12-14T03:56:07.311643797Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777376
[2023-12-14T03:56:07.311645420Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777377
[2023-12-14T03:56:07.311647153Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777378
[2023-12-14T03:56:07.311651251Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777379
[2023-12-14T03:56:07.311653265Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777380
[2023-12-14T03:56:07.311655549Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777382
[2023-12-14T03:56:07.311657663Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777383
[2023-12-14T03:56:07.311669766Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777384
[2023-12-14T03:56:07.311671599Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777385
[2023-12-14T03:56:07.311673243Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777386
[2023-12-14T03:56:07.311674866Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777387
[2023-12-14T03:56:07.311676629Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777388
[2023-12-14T03:56:07.311678252Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777389
[2023-12-14T03:56:07.311679925Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777390
[2023-12-14T03:56:07.311681639Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777391
[2023-12-14T03:56:07.311683252Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777392
[2023-12-14T03:56:07.311684845Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777393
[2023-12-14T03:56:07.311686458Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777394
[2023-12-14T03:56:07.311688061Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777395
[2023-12-14T03:56:07.311689744Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777396
[2023-12-14T03:56:07.311692038Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777397
[2023-12-14T03:56:07.311693682Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777398
[2023-12-14T03:56:07.311695816Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777399
[2023-12-14T03:56:07.311697429Z INFO  solana_runtime::bank_forks] Populating 235777401 as a descendant of 235777400
```

So, I think the code should stay (altho I removed the looping in https://github.com/solana-labs/solana/pull/34449/commits/9cac54c8c703d040dca3e524e4c46d346bcc3a66)